Sources/TokamakCore/Views/Selectors/Slider.swift

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Tokamak contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+public enum _SliderStep {
+  case any
+  case discrete(Double.Stride)
+}
+
+private func convert<T: BinaryFloatingPoint>(_ binding: Binding<T>) -> Binding<Double> {
+  Binding(get: { Double(binding.wrappedValue) }, set: { binding.wrappedValue = T($0) })
+}
+
+private func convert<T: BinaryFloatingPoint>(_ range: ClosedRange<T>) -> ClosedRange<Double> {
+  Double(range.lowerBound)...Double(range.upperBound)
+}
+
+/// A control for selecting a value from a bounded linear range of values.
+///
+/// Available when `Label` and `ValueLabel` conform to `View`.
+public struct Slider<Label, ValueLabel>: View where Label: View, ValueLabel: View {
+  let label: Label
+  let minValueLabel: ValueLabel
+  let maxValueLabel: ValueLabel
+  let valueBinding: Binding<Double>
+  let bounds: ClosedRange<Double>
+  let step: _SliderStep
+  // FIXME: IMPLEMENT “For example, on iOS, a Slider is considered to be actively
+  // editing while the user is touching the knob and sliding it around the track.”
+  let onEditingChanged: (Bool) -> ()
+
+  public var body: Never {
+    neverBody("Slider")
+  }
+}
+
+extension Slider where Label == EmptyView, ValueLabel == EmptyView {
+  public init<V>(
+    value: Binding<V>,
+    in bounds: ClosedRange<V> = 0...1,
+    onEditingChanged: @escaping (Bool) -> () = { _ in }
+  ) where V: BinaryFloatingPoint, V.Stride: BinaryFloatingPoint {
+    label = EmptyView()
+    minValueLabel = EmptyView()
+    maxValueLabel = EmptyView()
+    valueBinding = convert(value)
+    self.bounds = convert(bounds)
+    step = .any
+    self.onEditingChanged = onEditingChanged
+  }
+
+  public init<V>(
+    value: Binding<V>,
+    in bounds: ClosedRange<V>,
+    step: V.Stride = 1,
```

Comment:
I'll be able to try this in action later today, but just looking at the code, I'm not sure how this overloaded function is distinguished from the one above when the `step` argument is omitted and the default `1` value is supplied? The types of the rest of the arguments seem to be identical to me, but maybe I'm missing something? Is this how SwiftUI defines this as well?

Reply:
From SwiftUI `.swiftinterface`:
```swift
extension Slider {
  @available(tvOS, unavailable)
  public init<V>(value: SwiftUI.Binding<V>, in bounds: Swift.ClosedRange<V> = 0...1, onEditingChanged: @escaping (Swift.Bool) -> Swift.Void = { _ in }, minimumValueLabel: ValueLabel, maximumValueLabel: ValueLabel, @SwiftUI.ViewBuilder label: () -> Label) where V : Swift.BinaryFloatingPoint, V.Stride : Swift.BinaryFloatingPoint
  public init<V>(value: SwiftUI.Binding<V>, in bounds: Swift.ClosedRange<V>, step: V.Stride = 1, onEditingChanged: @escaping (Swift.Bool) -> Swift.Void = { _ in }, minimumValueLabel: ValueLabel, maximumValueLabel: ValueLabel, @SwiftUI.ViewBuilder label: () -> Label) where V : Swift.BinaryFloatingPoint, V.Stride : Swift.BinaryFloatingPoint
}
```
I agree that it seems ambiguous though.

Reply:
If you specify `step`, you must specify `bounds` (ie `Slider(value: $foo, step: 2)` is invalid). Not sure how it deduplicates, though.

Maybe the reason is that tvOS doesn’t support the continuous version so they have the overload so a `Slider(value: $foo, in: 0...100)` for example works on both platforms?